lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -1405,6 +1463,23 @@ unsafe fn from_go_array<T: Clone>(data: *mut T, len: u32) -> Vec<T> {
     slice::from_raw_parts(data, len as usize).to_vec()
 }
 
+/// to_c_string can be used to return string values over the Go boundary.
+/// To avoid memory leaks, the Go function must call free_go_string once
+/// it's done with the memory.
+///
+/// See https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw
+fn to_c_string(s: &str) -> Result<*const c_char, NulError> {
```

Comment:
I tried initially to stick with that pattern, however what I realized is that it only works in cases where we pass a string from Rust to Go by _calling a Go function from Rust_. What we're trying to do here is pass a string from Rust to Go by _calling a Rust function from Go_.

For example we pass a string from Rust to Go here when we call `let err = tdp_sd_info_request(...`
https://github.com/gravitational/teleport/blob/f49564cf1e0b370b3789a8482796c1f01dbcc9c0/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L332-L362
The underlying memory of `path: c_string.as_ptr(),` stays valid throughout that call, because the Rust function we're calling from hasn't returned yet.

However when I tried to do the same sort of thing from `get_server_disconnect_reason`, I would end up with gibberish for the string on the Go side.

```rust
pub unsafe extern "C" fn get_server_disconnect_reason(client_ptr: *mut Client) -> ReasonOrError {
// ...
    let reason = rdp_client.global.get_server_disconnect_reason();
    match CString::new(reason) {
        Ok(c_string) => {
            return ReasonOrError {
                reason: c_string.as_ptr(), // ends up as gibberish when I C.GoString() it on the Go side
                err: CGOErrCode::ErrCodeSuccess,
            }
        }
        Err(e) => {
            // ...
        }
    }
}
```

It makes sense if you look closer with Rust's memory semantics in mind: `c_string` is owned by the `get_server_disconnect_reason` function, so as soon as that function returns, that memory is invalidated. So we call it on the Go side, it returns, and then we try to `C.GoString` invalid memory.

We used to have functions like these back when we were passing strings from Rust to Go as errors. Now we typically just pass error codes which are stack allocated `uint32`, so we got rid of them.

Thinking about it more though, I think I can get us back to each side being fully responsible for freeing whatever memory it allocates by changing the API to this function so that Go passes it a buffer that it copies into.

Reply:
Ah, of course. Thanks for the explanation.

Reply:
See b6032a5797a074cfa2c22b12cb0ed587e0c3fcd7

The API is messier now, but it sticks to the principle of each side being responsible for free-ing whatever memory it allocates.

I'm skeptical that this is actually better. The API is clearly worse and the internals that implement it are uglier and more complex. The original principle was conjured up in partial ignorance of the problem described above. Additionally it's not stuck to everywhere, for example the `Client` itself is initialized by Rust, and then Go is later responsible for calling `free_rdp` to free it.

Perhaps we should revise it to

1. When we pass heap allocated memory from Rust to Go by passing it into a Go function called from Rust, or vise versa, the side that allocated the memory is responsible for freeing it, and the side that received the memory is responsible for immediately copying it if it's going to be needed after the function returns.
2. When we pass heap allocated memory from Rust to Go by returning it from a Rust function called by Go, or vise versa, the memory allocator is responsible for providing another function to free the heap allocated memory, and the caller is responsible for calling that function once it's done with the heap allocated memory.

Reply:
I agree, doesn't seem like much of an improvement. I'm fine with the original approach or your revised suggestion.